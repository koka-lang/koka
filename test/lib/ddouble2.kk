module ddouble2

public import std/num/ddouble
public import std/num/double

fun check(name : string, res : string, tst : () -> io string ) : io () {
  val got = tst()
  println(name.pad-right(8,' ') + ": "
    + (if (got == res) then "ok: " + res
                       else "FAILED!:\n expect: " + res + "\n gotten: " + got + "\n"))
  if (got != res) then error("failed") else ()
}

fun check-rt( name : string, expect : ddouble, prec : int, tst : () -> io ddouble ) : io () {
  check(name,expect.show(prec)) {
    check-round-trip(name,tst()).show(prec)
  }
}

fun check-rte( name : string, expect : string, tst : () -> io ddouble ) : io () {
  check(name,expect) {
    check-round-trip(name,tst()).show-exp
  }
}

fun check-round-trip( name : string, x : ddouble ) : io ddouble {
  //val y = x.show(33).parse-ddouble.default(dd-nan)
  val y = x.show-hex.parse-ddouble.default(dd-nan)
  if (y==x) then () else {
      println("failed round-trip parse! (" + name + ")\n x : " +
            x.show-exp + "\n     " + x.show-hex +  "\n y : " + y.show-exp + "\n     " + y.show-hex)
  }
  x
}

fun check-prec( name : string, expect : ddouble, res : ddouble, eps : ddouble = ddouble-exp(1,-15) ) : io () {
  val pre = name.pad-right(8,' ') + ": "
  if (expect.finite?) then {
    val delta  = (res - expect).abs
    if (delta < eps) then {
      println(pre + "ok: " + res.show + ", dt: " + delta.show(6)  )
    }
    else {
      val relerr = (delta / expect).abs
      val releps = if (expect.abs < ten) then eps else eps * 1000.ddouble
      if (relerr < releps) then {
        println(pre + "ok: " + res.show + ", derr: " + relerr.show(6)  )
      }
      else {
        println(pre + "FAILED:\n expect: " + expect.show + "\n gotten: " + res.show )
        println("delta : " + delta.show + "\n error: " + relerr.show + "\n eps   : " + releps.show )
        error("failed " + name)
      }
    }
  }
  else {
    if (expect == res) then {
      println( pre + "ok: " + res.show )
    }
    else {
      println(pre + "FAILED:\n expect: " + expect.show + "\n gotten: " + res.show )
      error("failed " + name)
    }
  }
}


public fun main() {
  test-e()
  test-pi-machin()
  test-pi-borwein()
  test-pi-salmanbrent()
  test-log2()
  test-exp()
  test-sincos()
}


/*-------------------------------------------------------
  Various formulas to approximate known numbers
-------------------------------------------------------*/

public fun test-e(){
  /* Use Taylor series
       e = 1 + 1 + 1/2! + 1/3! + 1/4! + ...
     To compute e.
  */
  check-rt("e",dd-e,31) {
      val approx = expand-e()
      val delta  = (approx - dd-e).abs
      val err    = delta / dd-epsilon
      //trace(" delta: " + delta.show + " = " + (delta / dd-epsilon).show(3) + " eps")
      //trace(" e: " + dd-e.show)
      // (approx.show + ", err: " + err.show(3) + " ulp")
      approx
  }
}

fun expand-e( approx : ddouble = 2.0.ddouble, t0 : ddouble = one, n0 : ddouble = one ) : div ddouble  {
  if (t0 <= dd-epsilon) then approx else {
    val n = n0 + one
    val t = t0 / n
    expand-e( approx + t, t, n )
  }
}


public fun test-log2() {
  check-rt("log2",dd-log2,31) {
    val approx = log2()
    val delta  = (approx - dd-log2).abs
    val err    = delta / dd-epsilon
    //trace(" delta: " + delta.show + " = " + (delta / dd-epsilon).show(3) + " eps")
    //trace(" log2: " + dd-log2.show )
    //(approx.show + ", err: " + err.show(3) + " ulp")
    approx
  }
}

fun log2() : div ddouble {
  // Use the Taylor series
  //  -log(1-x) = x + x^2/2 + x^3/3 + x^4/4 + ...
  // with x = 1/2 to get  log(1/2) = -log 2.
  expand-log2(0.5.ddouble,0.5.ddouble)
}

fun expand-log2( acc : ddouble, t : ddouble, n : ddouble  = one ) : div ddouble {
  if (t <= dd-epsilon) then acc else {
    val n1 = n + one
    val t1 = t * 0.5.ddouble
    val acc1 = acc + (t1 / n1)
    expand-log2( acc1, t1, n1 )
  }
}


// check exponent equalities:
//   e^2 = exp(2)
//       = exp(-13/4) * exp(-9/4) * exp(-5/4) * exp(-1/4) *
//          exp(3/4) * exp(7/4) * exp(11/4) * exp(15/4)
public fun test-exp() {
  val expect = (dd-e * dd-e)

  check-rt("exp1",expect,31) { dd-e.sqr }
  check-rt("exp2",expect,30) { exp(2.ddouble) }
  check-rt("exp3",expect,30) {
    list(-3,4).map( fun(i) {
      exp((4*i).dec.ddouble / 4.ddouble)
    }).foldl(one,(*))
  }
}

/* Check sin/cos equalities:
  sin(x) = sin(5x/7)cos(2x/7) + cos(5x/7)sin(2x/7)
  cos(x) = cos(5x/7)cos(2x/7) - sin(5x/7)sin(2x/7)
*/
public fun test-sincos() {
  [1,2,4,3,5,6,7,8,9,10,1000,10000000000]
    .foreach-indexed(fun(i,r) {
      check-sincos(i, dd-twopi / r.ddouble)
      check-sincos(i, ~dd-twopi / r.ddouble)
    })
}

public fun check-sincos( i : int, x : ddouble ) {
  val x1 = (5.ddouble * x) / 7.ddouble
  val x2 = (2.ddouble * x) / 7.ddouble

  val s1 = sin(x)
  val c1 = cos(x)
  val s2 = sin(x1) * cos(x2) + cos(x1) * sin(x2)
  val c2 = cos(x1) * cos(x2) - sin(x1) * sin(x2)
  check-prec("sin(" + x.show + ")", s1, s2, ddouble-exp(1,-31))  
  check-prec("cos(" + x.show + ")", c1, c2, ddouble-exp(1,-31))

  // check not too much error with respect to the double computation
  check-prec( "dsin" + i.show, sin(x.double).ddouble, s1, ddouble-exp(1,-15) )  
  check-prec( "dcos" + i.show, cos(x.double).ddouble, c1, ddouble-exp(1,-15) )
}

/*-------------------------------------------------------
  Various formulas to approximate pi
-------------------------------------------------------*/

//-----------------------------------------------
// Use the Machin's arctangent formula:
//    pi / 4  =  4 arctan(1/5) - arctan(1/239)
// The arctangent is computed based on the Taylor series expansion
//   arctan(x) = x - x^3 / 3 + x^5 / 5 - x^7 / 7 + ...
public fun test-pi-machin() {
  check-rt("pi-mach", dd-pi,29) {
    val approx = 4.ddouble * ((4.ddouble * arctan(one / 5.ddouble )) - arctan( one / 239.ddouble ))
    val delta  = (approx - dd-pi).abs
    val err    = delta / dd-epsilon
    //trace(" delta: " + delta.show + " = " + (delta / dd-epsilon).show(3) + " eps")
    //trace(" pi: " + dd-pi.show )
    // (approx.show + ", err: " + err.show(3) + " ulp")
    approx
  }
}

public fun arctan( x : ddouble ) : div ddouble {
  expand-arctan(x,x.sqr)
}

fun expand-arctan( t : ddouble, xsqr : ddouble, n : ddouble = one, acc : ddouble = zero, pos : bool = True ) : div ddouble {
  if (t <= dd-epsilon) then acc else {
    val acc1 = if (pos) then acc + (t / n) else acc - (t / n)
    val n1 = n + 2.ddouble
    val t1 = t * xsqr
    expand-arctan( t1, xsqr, n1, acc1, !pos )
  }
}


// --------------------------------
// Borwein Quartic algorithm for pi
//
public fun test-pi-borwein() {
  check-rt("borwein", dd-pi, 29) {
    val a = 6.ddouble - (4.ddouble * sqrt(2.ddouble))
    val y = sqrt(2.ddouble) - one
    val m = 2.ddouble
    val p = one / a
    val approx = expand-pi-borwein( a, y, m, p )
    approx
  }
}

fun expand-pi-borwein( a : ddouble, y : ddouble, m : ddouble,
                       p : ddouble, i : int = 0 ) : div ddouble
{
  val m1 = m * 4.ddouble
  val r  = nroot( one - y.sqr.sqr, 4 )
  val y1 = (one - r) / (one + r)
  val a1 = (a * sqr(sqr(one + y1))) - (m1*y1*(one + y1 + y1.sqr))

  val p1 = one / a1
  if (i >= 20 || ((p1 - p).abs < 4.ddouble * dd-epsilon))
   then p1
   else expand-pi-borwein(a1,y1,m1,p1,i.inc)
}


// --------------------------------
// Salman Brent approximation to pi
//
public fun test-pi-salmanbrent() {
  check("pi-sb",dd-pi.show(29)) {
    val a = one
    val b = sqrt(0.5.ddouble)
    val s = 0.5.ddouble
    val m = one
    val p = (2.0.ddouble * sqr(a)) / s
    val approx = expand-pi-sb(a,b,s,m,p)
    approx.show(29)
  }
}

public fun showx( x : ddouble ) : string {
  x.show-exp(dd-max-prec+1)
}

fun expand-pi-sb( a : ddouble, b : ddouble, s : ddouble,
                  m : ddouble, p : ddouble, i : int = 0 ) : div ddouble
{
  val m1 = m * 2.0.ddouble
  val a1 = 0.5.ddouble * (a + b)
  val b1 = a * b
  val s1 = s - (m1 * (a1.sqr - b1))
  val b2 = sqrt(b1)
  //trace(" a: " + a.showx + ", b: " + b.showx)
  //trace(" b2: " + b2.showx + ", b1: " + b1.showx )
  val p1 = (2.0.ddouble * a1.sqr) / s1
  //trace("i: " + i.show + ": p: " + p1.show-exp(dd-max-prec+1))
  if (i >= 20 || ((p1 - p).abs <= 64.ddouble*dd-epsilon))
   then p1
   else expand-pi-sb(a1,b2,s1,m1,p1,i.inc)
}
