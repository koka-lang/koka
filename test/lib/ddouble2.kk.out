e : ok: 2.7182818284590452353602874713527
pi-mach : ok: 3.14159265358979323846264338328
borwein : ok: 3.14159265358979323846264338328
pi-sb : ok: 3.14159265358979323846264338328
log2 : ok: 0.6931471805599453094172321214582
exp1 : ok: 7.3890560989306502272304274605750
exp2 : ok: 7.389056098930650227230427460575
exp3 : ok: 7.389056098930650227230427460575
sin(6.283185307179586476925286766559): ok: 0, dt: 0.000000
cos(6.283185307179586476925286766559): ok: 1, dt: 1.694818e-32
dsin0 : ok: 0, dt: 2.449294e-16
dcos0 : ok: 1, dt: 0.000000
sin(-6.283185307179586476925286766559): ok: 0, dt: 0.000000
cos(-6.283185307179586476925286766559): ok: 1, dt: 1.694818e-32
dsin0 : ok: 0, dt: 2.449294e-16
dcos0 : ok: 1, dt: 0.000000
sin(3.1415926535897932384626433832795): ok: 0, dt: 0.000000
cos(3.1415926535897932384626433832795): ok: -1, dt: 6.162976e-33
dsin1 : ok: 0, dt: 1.224647e-16
dcos1 : ok: -1, dt: 0.000000
sin(-3.1415926535897932384626433832795): ok: 0, dt: 0.000000
cos(-3.1415926535897932384626433832795): ok: -1, dt: 6.162976e-33
dsin1 : ok: 0, dt: 1.224647e-16
dcos1 : ok: -1, dt: 0.000000
sin(1.5707963267948966192313216916398): ok: 1, dt: 2.542228e-32
cos(1.5707963267948966192313216916398): ok: 0, dt: 0.000000
dsin2 : ok: 1, dt: 0.000000
dcos2 : ok: 0, dt: 6.123234e-17
sin(-1.5707963267948966192313216916398): ok: -1, dt: 2.542228e-32
cos(-1.5707963267948966192313216916398): ok: 0, dt: 0.000000
dsin2 : ok: -1, dt: 0.000000
dcos2 : ok: 0, dt: 6.123234e-17
sin(2.0943951023931954923084289221863): ok: 0.8660254037844386467637231707529, dt: 7.395571e-32
cos(2.0943951023931954923084289221863): ok: -0.5, dt: 7.626683e-32
dsin3 : ok: 0.866025403784438646763723170753, dt: 5.017542e-17
dcos3 : ok: -0.5, dt: 2.220446e-16
sin(-2.0943951023931954923084289221863): ok: -0.8660254037844386467637231707529, dt: 7.395571e-32
cos(-2.0943951023931954923084289221863): ok: -0.5, dt: 7.626683e-32
dsin3 : ok: -0.866025403784438646763723170753, dt: 5.017542e-17
dcos3 : ok: -0.5, dt: 2.220446e-16
sin(1.2566370614359172953850573533118): ok: 0.9510565162951535721164393333793, dt: 1.848893e-32
cos(1.2566370614359172953850573533118): ok: 0.3090169943749474241022934171828, dt: 6.162976e-33
dsin4 : ok: 0.9510565162951535721164393333794, dt: 4.093450e-17
dcos4 : ok: 0.3090169943749474241022934171829, dt: 2.716058e-17
sin(-1.2566370614359172953850573533118): ok: -0.9510565162951535721164393333793, dt: 1.848893e-32
cos(-1.2566370614359172953850573533118): ok: 0.3090169943749474241022934171828, dt: 6.162976e-33
dsin4 : ok: -0.9510565162951535721164393333794, dt: 4.093450e-17
dcos4 : ok: 0.3090169943749474241022934171829, dt: 2.716058e-17
sin(1.0471975511965977461542144610932): ok: 0.866025403784438646763723170753, dt: 0.000000
cos(1.0471975511965977461542144610932): ok: 0.5, dt: 1.694818e-32
dsin5 : ok: 0.866025403784438646763723170753, dt: 6.084688e-17
dcos5 : ok: 0.5, dt: 1.110223e-16
sin(-1.0471975511965977461542144610932): ok: -0.866025403784438646763723170753, dt: 0.000000
cos(-1.0471975511965977461542144610932): ok: 0.5, dt: 1.694818e-32
dsin5 : ok: -0.866025403784438646763723170753, dt: 6.084688e-17
dcos5 : ok: 0.5, dt: 1.110223e-16
sin(0.897597901025655210989326680937): ok: 0.7818314824680298087084445266741, dt: 3.928897e-32
cos(0.897597901025655210989326680937): ok: 0.6234898018587335305250048840042, dt: 9.860761e-32
dsin6 : ok: 0.781831482468029808708444526674, dt: 5.074320e-18
dcos6 : ok: 0.6234898018587335305250048840043, dt: 6.386130e-17
sin(-0.897597901025655210989326680937): ok: -0.7818314824680298087084445266741, dt: 3.928897e-32
cos(-0.897597901025655210989326680937): ok: 0.6234898018587335305250048840042, dt: 9.860761e-32
dsin6 : ok: -0.781831482468029808708444526674, dt: 5.074320e-18
dcos6 : ok: 0.6234898018587335305250048840043, dt: 6.386130e-17
sin(0.7853981633974483096156608458199): ok: 0.7071067811865475244008443621049, dt: 0.000000
cos(0.7853981633974483096156608458199): ok: 0.7071067811865475244008443621049, dt: 1.848893e-32
dsin7 : ok: 0.7071067811865475244008443621049, dt: 6.268584e-17
dcos7 : ok: 0.7071067811865475244008443621049, dt: 4.833647e-17
sin(-0.7853981633974483096156608458199): ok: -0.7071067811865475244008443621049, dt: 0.000000
cos(-0.7853981633974483096156608458199): ok: 0.7071067811865475244008443621049, dt: 1.848893e-32
dsin7 : ok: -0.7071067811865475244008443621049, dt: 6.268584e-17
dcos7 : ok: 0.7071067811865475244008443621049, dt: 4.833647e-17
sin(0.6981317007977318307694763073954): ok: 0.6427876096865393263226434099072, dt: 1.232595e-32
cos(0.6981317007977318307694763073954): ok: 0.7660444431189780352023926505555, dt: 9.244464e-33
dsin8 : ok: 0.6427876096865393263226434099072, dt: 7.442622e-17
dcos8 : ok: 0.7660444431189780352023926505555, dt: 2.175071e-17
sin(-0.6981317007977318307694763073954): ok: -0.6427876096865393263226434099072, dt: 1.232595e-32
cos(-0.6981317007977318307694763073954): ok: 0.7660444431189780352023926505555, dt: 9.244464e-33
dsin8 : ok: -0.6427876096865393263226434099072, dt: 7.442622e-17
dcos8 : ok: 0.7660444431189780352023926505555, dt: 2.175071e-17
sin(0.6283185307179586476925286766559): ok: 0.5877852522924731291687059546391, dt: 5.084455e-32
cos(0.6283185307179586476925286766559): ok: 0.8090169943749474241022934171828, dt: 1.232595e-32
dsin9 : ok: 0.5877852522924731291687059546391, dt: 7.934751e-18
dcos9 : ok: 0.8090169943749474241022934171828, dt: 2.716058e-17
sin(-0.6283185307179586476925286766559): ok: -0.5877852522924731291687059546391, dt: 5.084455e-32
cos(-0.6283185307179586476925286766559): ok: 0.8090169943749474241022934171828, dt: 1.232595e-32
dsin9 : ok: -0.5877852522924731291687059546391, dt: 7.934751e-18
dcos9 : ok: 0.8090169943749474241022934171828, dt: 2.716058e-17
sin(0.0062831853071795864769252867666): ok: 0.006283143965558951249735310158, dt: 9.148167e-34
cos(0.0062831853071795864769252867666): ok: 0.9999802608561371298486736604302, dt: 3.081488e-33
dsin10 : ok: 0.006283143965558951249735310158, dt: 1.654749e-19
dcos10 : ok: 0.9999802608561371298486736604302, dt: 1.451583e-17
sin(-0.0062831853071795864769252867666): ok: -0.006283143965558951249735310158, dt: 9.148167e-34
cos(-0.0062831853071795864769252867666): ok: 0.9999802608561371298486736604302, dt: 3.081488e-33
dsin10 : ok: -0.006283143965558951249735310158, dt: 1.654749e-19
dcos10 : ok: 0.9999802608561371298486736604302, dt: 1.451583e-17
sin(6.283185307179586476925286766559e-10): ok: 6.2831853071795864765118697441513e-10, dt: 5.739719e-42
cos(6.283185307179586476925286766559e-10): ok: 0.9999999999999999998026079119782, dt: 0.000000
dsin11 : ok: 6.2831853071795864765118697441512e-10, dt: 2.803512e-26
dcos11 : ok: 0.9999999999999999998026079119782, dt: 1.973921e-19
sin(-6.283185307179586476925286766559e-10): ok: -6.2831853071795864765118697441513e-10, dt: 5.739719e-42
cos(-6.283185307179586476925286766559e-10): ok: 0.9999999999999999998026079119782, dt: 0.000000
dsin11 : ok: -6.2831853071795864765118697441512e-10, dt: 2.803512e-26
dcos11 : ok: 0.9999999999999999998026079119782, dt: 1.973921e-19